go/tools/builders/compilepkg.go

```diff
@@ -270,16 +271,25 @@ func compileArchive(
 	// constraints.
 	compilingWithCgo := haveCgo && cgoEnabled
 
+	filterForNogo := func(slice []string) []string {
+		var filtered []string
+		for _, s := range slice {
+			if !ignoreForNogo[s] {
+				filtered = append(filtered, s)
+			}
+		}
+		return filtered
+	}
 	// When coverage is set, source files will be modified during instrumentation. We should only run static analysis
 	// over original source files and not the modified ones.
 	// goSrcsNogo and cgoSrcsNogo are copies of the original source files for nogo to run static analysis.
-	goSrcsNogo := goSrcs
+	goSrcsNogo := filterForNogo(goSrcs)
 	cgoSrcsNogo := cgoSrcs
 
 	// Instrument source files for coverage.
 	if coverMode != "" {
 		// deep copy original source files for nogo static analysis, avoid being modified by coverage.
-		goSrcsNogo = append([]string{}, goSrcs...)
+		goSrcsNogo = filterForNogo(append([]string{}, goSrcs...))
```

Comment:
We already filter once outside the `if` clause, let’s just re-use the result?

```suggestion
		goSrcsNogo = filterForNogo(append([]string{}, goSrcsNogo...))
```

Reply:
and since the filterForNogo is creating a copied slice anyway, I think we could simplify this further.

Reply:
+1, `filterForNogo` is effectively a deep copy

Reply:
Changed this by making both deep copies unconditional, the alternative just seemed too confusing for little gain.